File: src/models/message.js
/****************************************************************
 * IMPORTS
 ****************************************************************/

 const {
    CENSOR_ATTRIBUTE,
    censorMessage,
} = require('./../core/logging.js');
const { User } = require('./users.js');

/****************************************************************
 * Class Message
 ****************************************************************/

class Message {
    text;
    lang;
    is_bot;
    chatId;
    userId;
    messageId;

    constructor(msg) {
        const date = (msg || {}).date;
        this.timestamp = typeof date === 'number' ? date * 1000 : undefined;
        this.text = ((msg || {}).text || '').trim();
        this.lang = ((msg || {}).from || {}).language_code;
        this.is_bot = (((msg || {}).from) || {}).is_bot;
        // IDs:
        this.messageId = (msg || {}).message_id;
        this.chatId = ((msg || {}).chat || {}).id;
        this.userId = ((msg || {}).from || {}).id;
    }

    getChatId() { return this.chatId; }

    getMessageId() { return this.messageId; }

    getUserId() { return this.userId; }

    getIds() { return { chatId: this.chatId, userId: this.userId, messageId: this.messageId }; }

    getText() { return this.text; }

    getLanguage() { return this.lang; }

    isBot() { return this.is_bot; }

    /********
     * Returns timestamp in ms
     ********/
    getTimestamp() { return this.timestamp; }

    /********
     * Returns User class, if data can be retrieved or else undefined.
     ********/
    async getUser(bot) {
        return bot.telegram.getChatMember(this.chatId, this.userId)
            .then((data) => new User(data))
            .catch((_) => {return undefined});
    }

    messageTooOld(t, expiry) {
        return (typeof this.timestamp === 'number') ? (this.timestamp + expiry < t) : true;
    }

    toRepr() {
        return {
            timestamp: this.timestamp,
            text:      this.text,
            lang:      this.lang,
            is_bot:    this.is_bot,
            messageId: this.messageId,
            chatId:    this.chatId,
            userId:    this.userId,
        }
    }

    toString() { return JSON.stringify(this.toRepr()); }

    /********
     * Provides a censored representation of Message:
     * - text` content of message partially censored.
     * - fully censored, if `full_censor=true` passed as argument.
     ********/
    toCensoredRepr(full_censor=false) {
        return {
            timestamp: this.timestamp,
            text:      full_censor === false ? censorMessage(this.text) : CENSOR_ATTRIBUTE,
            lang:      this.lang,
            is_bot:    this.is_bot,
            messageId: this.messageId,
            chatId:    this.chatId,
            userId:    this.userId,
        }
    }

    /********
     * Provides a censored representation of Message:
     * - text` content of message partially censored.
     * - fully censored, if `full_censor=true` passed as argument.
     ********/
    toCensoredString(full_censor) { return JSON.stringify(this.toCensoredRepr(full_censor)); }
}

/****************************************************************
 * EXPORTS
 ****************************************************************/

module.exports = {
    Message
};

File: src/models/telegram/message.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# IMPORTS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

from src.thirdparty.api import *;
from src.thirdparty.code import *;
from src.thirdparty.config import *;
from src.thirdparty.misc import *;
from src.thirdparty.types import *;

from src.core.utils import *;
from src.core.log_special import *;
from src.models.telegram.bot import *;
from src.models.telegram.users import *;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# EXPORTS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

__all__ = [
    'Message',
];

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# CLASS Message
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

class Message():
    text:      str;
    lang:      str;
    is_bot:    str;
    chatId:    int;
    userId:    int;
    messageId: int;
    timestamp: datetime;

    def __init__(self, msg: TgMessage):
        self.timestamp = datetime.fromtimestamp(msg.date);
        self.text = msg.text.strip();
        self.lang = msg.from_user.language_code;
        self.is_bot = msg.from_user.is_bot;
        # IDs:
        self.messageId = msg.message_id;
        self.chatId = msg.chat.id;
        self.userId = msg.from_user.id;

    def getChatId(self) -> int:
        return self.chatId;

    def getMessageId(self) -> int:
        return self.messageId;

    def getUserId(self) -> int:
        return self.userId;

    def getText(self) -> str:
        return self.text;

    def getLanguage(self) -> str:
        return self.lang;

    def isBot(self) -> bool:
        return self.is_bot;

    def getTimestamp(self) -> datetime:
        '''
        Returns timestamp in ms
        '''
        return self.timestamp;

    @wrap_output_as_option
    def getUser(self, bot: MyBot) -> User:
        '''
        Returns User class,
        if data can be retrieved or else undefined.
        '''
        return User(bot.get_chat_member(chat_id=self.chatId, user_id=self.userId));

    def messageTooOld(self, now: datetime, expiry: timedelta) -> bool:
        '''
        NOTE:
        - `self.timestamp` = time at which message was posted.
        - `now` = time at which command is processed

        @returns
        True if and only if `now` is later than 'expiration' of post
        (computed as `self.timestamp + expiry`).
        '''
        return self.timestamp + expiry < now;

    def toRepr(self) -> Dict[str, Any]:
        return {
            'timestamp': str(self.timestamp),
            'text':      self.text,
            'lang':      self.lang,
            'is_bot':    self.is_bot,
            'messageId': self.messageId,
            'chatId':    self.chatId,
            'userId':    self.userId,
        };

    def __str__(self) -> str:
        return json.dumps(self.toRepr());

    def toCensoredRepr(self, full_censor: bool = False) -> Dict[str, Any]:
        '''
        Provides a censored representation of Message:
        - text` content of message partially censored.
        - fully censored, if `full_censor=true` passed as argument.
        '''
        return {
            'timestamp': str(self.timestamp),
            'text':      partiallyCensorMessage(self.text) if full_censor is False else CENSOR_ATTRIBUTE,
            'lang':      self.lang,
            'is_bot':    self.is_bot,
            'messageId': self.messageId,
            'chatId':    self.chatId,
            'userId':    self.userId,
        };

    def toCensoredString(self, full_censor: bool) -> str:
        '''
        Provides a censored representation of Message:
        - text` content of message partially censored.
        - fully censored, if `full_censor=true` passed as argument.
        '''
        return json.dumps(self.toCensoredRepr(full_censor));
